### PR TITLE
Initial Review

### PR DESCRIPTION
**Pull request by jtdoepke**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,40 @@
+sudo: required
```
Comment (chavlin) [inline .travis.yml]: How does this Travis file fit into the overall deployment of the service?  Can't find it referenced anywhere else in the code.  Does JSL use it?  Do any of our other services use Travis CI?

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+matrix:
+  allow_failures:
+  - go: tip
+
+env:
+  - GO111MODULE=on
+  - GO111MODULE=off
+
+addons:
+  # ssh_known_hosts: github.com
```
Comment (chavlin) [inline .travis.yml]: Why is this commented out at the moment?  What purpose did/will it serve?

File: Dockerfile.cloudwatcher [hunk 3]
```diff
@@ -0,0 +1,56 @@
+###############################
+# STAGE 1 test and build binary
+###############################
+
+FROM golang:1.12 AS builder
+
+# Set Go build env vars
+ARG GOOS=linux
```
Comment (chavlin) [inline Dockerfile.cloudwatcher]: Why declare these on their own line?  I don't think they're being used anywhere below

File: Dockerfile.cloudwatcher [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+COPY . .
+
+# Install/Update packages (after src COPY so it always happens)
+RUN apt-get update -y \
+    && apt-get upgrade -y \
+    && apt-get install -y ca-certificates tzdata \
+    && update-ca-certificates
+
+# Build the binary
+RUN CGO_ENABLED=0 go build -ldflags='-w -s' -o /go/bin/cloudwatcher ./cmd/cloudwatcher
```
Comment (chavlin) [inline Dockerfile.cloudwatcher]: What's the ldflags arg all about?  I see that 
`	-ldflags '[pattern=]arg list'
		arguments to pass on each go tool link invocation.
`
But where are they getting passed, exactly, and what do they signify?

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,6 @@
+COMPOSE := docker-compose
+GO := go
+GOTEST := $(GO) test -race -timeout 10s
+
+test:
+	@$(COMPOSE) up -d elasticsearch localstack && $(COMPOSE) run wait && $(GOTEST) ./...; ret=$$?; $(COMPOSE) down -v && exit $$ret
```
Comment (chavlin) [inline Makefile]: What's the story with the `$$?` ?
$$ refers to the current process id
$? refers to the last executed command id (and I think this is the default arg for `exit`)
But what's `$$?` ?

File: .travis.yml (hunk 2 above)
Comment (jtdoepke) [reply .travis.yml]: [Cargo-culted](https://github.com/olivere/elastic/blob/release-branch.v7/.travis.yml#L14). Probably not needed.

- [ ] Remove `ssh_known_hosts` from .travis.yml

File: .travis.yml (hunk 1 above)
Comment (jtdoepke) [reply .travis.yml]: Travis is for [running tests](https://travis-ci.org/mintel/elasticsearch-asg).

Docker images are [built](https://cloud.docker.com/u/mintel/repository/docker/mintel/elasticsearch-snapshooter) [by](https://cloud.docker.com/u/mintel/repository/docker/mintel/elasticsearch-lifecycler) [Docker](https://cloud.docker.com/u/mintel/repository/docker/mintel/elasticsearch-cloudwatcher) [Hub](https://cloud.docker.com/u/mintel/repository/docker/mintel/elasticsearch-healthcheck).

There is zero Jenkins involvement with this repository.

File: Dockerfile.cloudwatcher (hunk 3 above)
Comment (jtdoepke) [reply Dockerfile.cloudwatcher]: These are part of the [standard `go build` env vars](https://gist.github.com/asukakenji/f15ba7e588ac42795f421b48b8aede63). I'm being explicit in case this is built on Mac/Windows/whatever (`go build` defaults to whatever OS/Arch it's running on). This will, for example, let @alphachai build and run this on his MacBook.

File: Dockerfile.cloudwatcher (hunk 4 above)
Comment (jtdoepke) [reply Dockerfile.cloudwatcher]: These make for a smaller Docker image. `-w` builds the binary without debugger information, and `-s` builds without the Go symbol table.
https://stackoverflow.com/questions/22267189/what-does-the-w-flag-mean-when-passed-in-via-the-ldflags-option-to-the-go-comman

File: Makefile (hunk 5 above)
Comment (jtdoepke) [reply Makefile]: Make also use `$` for variables in Makefiles. `$$` is a literal `$`, so the bash subprocess evaluates `ret=$?`.

File: .travis.yml (hunk 1 above)
Comment (chavlin) [reply .travis.yml]: I guess my question is more "What triggers the running of Travis within this deployment's infrastructure"?  Where is the code that looks at this configuration file?

Comment (chavlin) [reply .travis.yml]: jtdoepke gave me a quick walk-through; the short answer is that Travis integrates with Github to run and monitor testing.

File: cmd/snapshooter/README.md [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+Flags:
+      --help                    Show context-sensitive help (also try
+                                --help-long and --help-man).
+  -v, --verbose                 Show debug logging.
+      --window=P1M=PT1H... ...  Snapshot frequency + TTL. May be set
+                                multiple times. ISO 8601 Duration string
+                                format. Example: `--window P1M=PT1H` ==
+                                keep hourly snapshots for 1 month.
+  -d, --delete                  If set, clean up old snapshots. This is
+                                false be default for safety's sake.
```
Comment (chavlin) [inline cmd/snapshooter/README.md]: "false be default" => "false on default"

File: cmd/snapshooter/snap.go [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+// NewSnapshotWindow returns a new SnapshotWindow by parsing two ISO 8601 Duration strings.
+//
+// It returns an error if the duration strings cannot be parsed.
+func NewSnapshotWindow(every, keepFor string) (SnapshotWindow, error) {
+	sw := SnapshotWindow{}
+
+	d, err := ptime.ParseISO8601D(every)
+	if err != nil {
+		return SnapshotWindow{}, err
```
Comment (chavlin) [inline cmd/snapshooter/snap.go]: Why not return sw since that's already been created?

File: pkg/time/multiple.go [hunk 8]
```diff
@@ -0,0 +1,109 @@
+package time
+
+import (
+	"errors"
+	"sync"
+	"time"
+)
+
+// Ceil returns the result of rounding t up to a multiple of d (since the zero time).
```
Comment (chavlin) [inline pkg/time/multiple.go]: Based on time module documentation, it looks like Truncate rounds down and Round is what rounds up; not sure the behavior you want:
https://golang.org/pkg/time/#Time.Truncate

Comment (jtdoepke) [reply pkg/time/multiple.go]: `time.Round()` rounds up or down to the nearest. I wanted a func that always goes up.

File: cmd/snapshooter/snap.go (hunk 7 above)
Comment (jtdoepke) [reply cmd/snapshooter/snap.go]: Good point.

File: pkg/time/multiple.go (hunk 8 above)
Comment (chavlin) [reply pkg/time/multiple.go]: Nevermind, I see now that I completely neglected the Add() part of this function.

File: cmd/healthcheck/main.go [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+)
+
+func main() {
+	kingpin.CommandLine.Help = "Handle AWS Autoscaling Group Lifecycle hook events for Elasticsearch from an SQS queue."
+	kingpin.Parse()
+
+	logger := cmd.SetupLogging()
+	defer func() {
+		defer func() {
+			// Make sure any buffered logs get flushed before exiting successfully.
```
Comment (aspage2) [inline cmd/healthcheck/main.go]: n00B question: why is flushing the log buffer a good idea?

Comment (aspage2) [reply cmd/healthcheck/main.go]: *reads two lines below*

never mind :D

File: pkg/time/multiple.go (hunk 8 above)
Comment (chavlin) [reply pkg/time/multiple.go]: I actually want to double-down on my initial comment here; I agree that Ceil should always round up, but you get behavior like this when you supply certain values:

```
func main() {
	now := time.Now()
	d, _ := time.ParseDuration("50m")
	fmt.Println(d)
	fmt.Println(now)
	fmt.Println(now.Add(d))
	fmt.Println(now.Add(d).Truncate(d))
}

```
Returns this:
```
50m0s
2009-11-10 23:00:00 +0000 UTC m=+0.000000001
2009-11-10 23:50:00 +0000 UTC m=+3000.000000001
2009-11-10 23:20:00 +0000 UTC
```
The Truncate forces the final time to be earlier than the Added time (because 23:20 is 1400 minutes, the rounded-down multiple of 50).  Is this expected behavior?  If so we should change the name from Ceil to something else -- or if we only want to accept multiples of 10 or discrete units as Duration arguments, we should build that into the function.  

We should also probably write some tests for this function as well.  I'd be happy to try my hand at it once we've established whether the function is behaving as expected/needed.

File: pkg/time/parse.go [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+	// Month duration
+	Month = (time.Duration(30.436875*float64(Day)) / time.Second) * time.Second // truncate to second
+
+	// Year duration
+	Year = (time.Duration(365.2425*float64(Day)) / time.Second) * time.Second // truncate to second
+)
+
+const iso8601Group = `(?:(?P<%s>-?\d+(?:[,.]\d+)?)%s)?`
+
+var iso8601Duation = regexp.MustCompile(fmt.Sprintf(`^P(?:0|%s|%s)$`,
```
Comment (chavlin) [inline pkg/time/parse.go]: Duation => Duration

Comment (jtdoepke) [reply pkg/time/parse.go]: Whoops...

File: pkg/time/parse.go [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+	if err != nil {
+		panic(err)
+	}
+	return d
+}
+
+// addDurationMul returns d + (n * u), and a bool indicating there was no int64 overflow/underflow.
+func addDurationMul(d time.Duration, n float64, u time.Duration) (time.Duration, bool) {
+	r := n * float64(u)
+	if r > float64(math.MaxInt64) {
```
Comment (chavlin) [inline pkg/time/parse.go]: I don't know how robust this function is around the edge cases -- apparently math.MaxInt64 causes overflow when you cast it from float64 back to int64?  We should definitely write some tests for this one that explicitly test the edge cases...
```
func main() {
	maxIntFloat := float64(math.MaxInt64)
	backToInt := int64(maxIntFloat)
	fmt.Println(backToInt)
}
```
returns:
```
`-9223372036854775808`

File: pkg/time/multiple.go (hunk 8 above)
Comment (jtdoepke) [reply pkg/time/multiple.go]: That is the expected behaviour, yes, although not because 23:20 is 1400 minutes. `time.Time.Truncate()`, `time.Time.Round()`, and `Ceil()` all operate relative to the Zero Time (January 1, year 1, 00:00:00.000000000 UTC), so `2009-11-10 23:20:00 UTC` is [some multiple](https://www.wolframalpha.com/input/?i=(0001-1-1T00:00:00+UTC+to+2009-11-10T23:20:00+UTC)+%2F+50+minutes) of 50 minutes from that time. Check out the [ruminations in the `time` src about computations on time](https://golang.org/src/time/time.go#L346).

I spent two days thinking about how to do semantically-nice time multiples for the snapshots before giving up. The most important behaviour of `Ceil()` is that it's deterministic i.e. the snapshooter will always choose the same snapshot times no matter when it was started.

File: pkg/time/comp.go [hunk 12]
```diff
@@ -0,0 +1,8 @@
+package time
+
+import "time"
+
+// Between returns true if t is between start and end inclusive.
+func Between(t, start, end time.Time) bool {
```
Comment (chavlin) [inline pkg/time/comp.go]: I don't know if it's worth checking for this, but should we verify that `start` is before `end`?  (I.e. this would return True if the end time supplied was earlier than the start time.)

File: pkg/time/parse.go (hunk 11 above)
Comment (jtdoepke) [reply pkg/time/parse.go]: Yeah, I've got a refactor that will use [`math/big`](https://golang.org/pkg/math/big/) to handle this precisely. (Bonus: we get to drop the dependency on `github.com/JohnCGriffin/overflow` .) Just need to write some tests...

File: pkg/time/comp.go (hunk 12 above)
Comment (jtdoepke) [reply pkg/time/comp.go]: Oh, yeah, that's a good point. I'd probably just make it so the order doesn't matter:
```go
// Between returns true if t is between a and b inclusive.
func Between(t, a, b time.Time) bool {
	if b.Before(a) {
		a, b = b, a
	}
	return (a.Equal(t) || a.Before(t)) && (b.Equal(t) || b.After(t))
}
```

File: pkg/time/parse.go (hunk 11 above)
Comment (jtdoepke) [reply pkg/time/parse.go]: PR #2

File: pkg/time/comp.go (hunk 12 above)
Comment (jtdoepke) [reply pkg/time/comp.go]: PR #3

File: cmd/snapshooter/main.go [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+		// Format the name of the new snapshot based of the scheduled time.
+		snapshotName := nextSnapshot.Format(SnapshotFormat)
+
+		// Reset the timer for the next snapshot.
+		nextSnapshot = snapshotSchedule.Next()
+		startSnapshot = time.After(time.Until(nextSnapshot))
+
+		// Create the new snapshot, waiting for it to complete.
+		logger.Info("creating snapshot " + snapshotName)
+		timeout := strconv.FormatInt(int64(time.Until(nextSnapshot).Minutes()), 10) + "m"
```
[the remark this answers is not in the document]
Comment (jtdoepke) [reply cmd/snapshooter/main.go]: Snapshot operations taking too long is going to be a consideration no matter how the snapshots are managed:

> Only one snapshot process can be executed in the cluster at any time. While snapshot of a particular shard is being created this shard cannot be moved to another node, which can interfere with rebalancing process and allocation filtering. Elasticsearch will only be able to move a shard to another node (according to the current allocation filtering settings and rebalancing algorithm) once the snapshot is finished. [[src]](https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-snapshots.html#_snapshot)

I expect we will need monitoring/alerting around how long our snapshot operations are taking, and to occationally tune the snapshot frequency. It's also possible that automatically deleting old snapshots will take too long, and we can only clean them up manually once in a while during a down time.

- [ ] TODO: Expose Prometheus metrics about snapshot operation duration.

File: cmd/snapshooter/main.go [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+		"P3M=P1W",
+		"P3Y=P1M",
+	).StringMap()
+	delete       = kingpin.Flag("delete", "If set, clean up old snapshots. This is false by default for safety's sake.").Short('d').Bool()
+	repoName     = kingpin.Flag("repo", "Name of the snapshot repository.").Default("backups").String()
+	repoType     = kingpin.Flag("type", "If set, create a repository of this type before creating snapshots. See also: '--settings'").String()
+	repoSettings = kingpin.Flag("settings", "Use these settings creating the snapshot repository. May be set multiple times. Example: `--type=s3 --settings bucket=my_bucket`").StringMap()
+)
+
+func main() {
```
[the remark this answers is not in the document]
Comment (jtdoepke) [reply cmd/snapshooter/main.go]: Two instances of snapshooter with the same configuration shouldn't be dangerous, since they will deterministically choose when to create and delete snapshots. There would be a race condition between the instances, but the resulting snapshots [should be the same](https://sequencediagram.org/index.html#initialData=MoOwhgDgzgFg9nALgUwE4AICCBaAfAUQBswpEBLAYymTFQpgC4BhVGldKcaeRdAJgAMARgCc2AQHZsfAGzYhQ8QKUAoUJFgIUGAEJ4iJclRp1GLNsg5dNvQaPFTZ8xQOWuVB0pWq16eddxaaOg6DPioqHCoDOgBNugA7mSIMOgpYLzgALaWYISsYAAmAJ7oyAAeZKRQKiBIlqhkAOYwvHAAZrHW8PW6YZVtIGURUR7EXsa+MP7dQRiYDADyANIMcTzo7WQgVTDIhSpAA).

In any case, we [won't run more than one](https://gitlab.com/mintel/everest/terraform/blob/elasticsearch/modules/elasticsearch/snapshooter.tf#L83).

A simple cron wouldn't allow for complex logic like downsampling snapshots over time, but if we don't need complex logic (see my response to your other comment) then I agree it would be better.

[the remark this answers is not in the document]
Comment (xiang-zhu) [reply cmd/snapshooter/main.go]: What's the downsampling snapshots? The way I'm thinking is the snapshooter could provide operation logic for snapshot creation and removal, but leave the schedule bit to cron.

File: cmd/snapshooter/main.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (jtdoepke) [reply cmd/snapshooter/main.go]: Right. I'll refactor that loop to die with an error if the snapshot takes too long. That way we'll get a signal that something's wrong and it won't create snapshots whose names don't match the time they were made.

- [ ] TODO: Refactor snapshooter loop to die if snapshot takes too long.

File: cmd/snapshooter/main.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (jtdoepke) [reply cmd/snapshooter/main.go]: The downsampling is the removal logic, keeping more recent snapshots and fewer older snapshots ([example](https://github.com/mintel/elasticsearch-asg/blob/b4c1bf19442192c8f024efa5540044f59eca4ad9/cmd/snapshooter/README.md#example)).

Correct me if I'm wrong, but I think you're saying:
1. Have a cron that triggers snapshooter.
2. Snapshooter creates a snapshot at the time it is triggered.
3. After the snapshot is finished, snapshooter selects older snapshots to remove and deletes them.

If snapshooter does logic about which snapshots to remove, then I'm not in favour of cron because:
- We would have knowledge of snapshot scheduling in two places (the cron schedule for creating snapshots and snapshooter's schedule for removing them).
- The logic for which snapshots to remove would become more complex because snapshooter does not know the cron's schedule for creating snapshots.
- It's adds a [lot of Terraform](https://github.com/dxw/terraform-aws-ecs-scheduled-task/blob/master/main.tf).
- We haven't used scheduled ECS tasks anywhere else.

However, if we don't automatically remove snapshots, then I'm in favour of cron and snapshooter could be replaced by something as simple as `curl -X PUT "http://elasticsearch:9200/_snapshot/backups/$( date +"%y-%m-%d-%H-%M-%S" )?wait_for_completion=true"`.

[the remark this answers is not in the document]
Comment (xiang-zhu) [reply cmd/snapshooter/main.go]: The way I'm thinking it could work is like:

As the equivalent of "P1M=PT1H" in the example

For example, an hourly cronjob to run, with an argument like
./snapshooter --keep-for 1M --period hourly

Which will trigger snapshooter to run the snapshot, then parse through the hourly backup to remove old snapshots.

It essentially is the same logic of one run of the for loop in your current code. The snapshooter code will not handle any scheduling, only take the information through argument.

The reason I'm thinking about this is the scheduling bit in current code make it impractical to run multiple instances (with different config) against the same ES. And there's no way to prevent that except you have to set up the deployment to make sure there's a single instance running.

The scheduling bit feels belong to a global context, instead of a utility-like container run on the side. An alternative direction which will make me feel better about keeping current snapshooter is to have a more centralized deployment setup for the whole system with mechanics to make sure the singularity of the snapshooter.

It's not a blocker issue for now, as I think the scheduler implemented here would work, but I feel it's a concern on the system design level

File: Dockerfile.cloudwatcher [hunk 15]
```diff
@@ -0,0 +1,56 @@
+###############################
+# STAGE 1 test and build binary
```
Comment (alphachai) [inline Dockerfile.cloudwatcher]: I just noticed this says _test_ and build. Which line is the testing on?

Comment (jtdoepke) [reply Dockerfile.cloudwatcher]: Oops; that's old. I was running the tests as part of the Docker build, but some of the tests rely on timing and break on the spotty build hardware that DockerHub provides. (That may no long be the case, and tests that rely on timing aren't great anyway.)

- [ ] TODO: Refactor tests that rely on timing.
- [ ] TODO: Check if tests work on DockerHub.

File: cmd/cloudwatcher/metrics.go [hunk 16]
```diff
@@ -0,0 +1,506 @@
...
+
+	if s.count == 0 { // Shortcut for no nodes Add()-ed.
+		return metrics
+	}
+
+	// Sum vCPU counts.
+	vcpuCount := sum(s.vcpuCounts...)
+
+	// Sum loads.
+	load1m := sum(s.load1m...)
```
Comment (alphachai) [inline cmd/cloudwatcher/metrics.go]: Why do you use variatic args instead of passing a slice? Is that just how sum() works?

File: cmd/cloudwatcher/metrics.go [hunk 17]
```diff
@@ -0,0 +1,506 @@
...
+				Value:             aws.Float64((1.0 - (sumAvailableFSBytes / sumMaxFSBytes)) * 100), // CloudWatch percents are int 0-100
+			},
+		)
+	}
+
+	return metrics
+}
+
+// sum returns a sum of d.
+func sum(d ...float64) float64 {
```
Comment (alphachai) [inline cmd/cloudwatcher/metrics.go]: Why do you use variatic args instead of passing a slice? Is there an advantage either way?

File: cmd/lifecycler/README.md [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+Once more nodes are added to the cluster, Elasticsearch will rebalance the shards to spread them evenly across the cluster.
+This will cause high CPU usage, which will trigger more autoscaling, which will cause high CPU, which will....
+
+AWS Autoscaling Groups provide [Lifecycle Hooks](https://docs.aws.amazon.com/autoscaling/ec2/userguide/lifecycle-hooks.html),
+which allow setup or teardown operations when an autoscaled instance is launching or terminating.
+The Autoscaling Group won't scale up/down again until the Lifecycle Hook event finishes i.e. its timeout expires.
+
+Lifecycler consumes Lifecycle Hook events from an SQS queue, and:
+
+1. If the instance is terminating, drains shards from the node and excludes if from master voting.
```
Comment (alphachai) [inline cmd/lifecycler/README.md]: if -> it

File: cmd/cloudwatcher/metrics.go (hunk 17 above)
Comment (jtdoepke) [reply cmd/cloudwatcher/metrics.go]: Variadic lets you use it for both slices (`sum(d...)`) and a static number of variables (`sum(a, b, c)`). It's all slices behind the scenes anyway.

File: cmd/lifecycler/main.go [hunk 19]
```diff
@@ -0,0 +1,319 @@
...
+	votingLock sync.Mutex
+
+	// votingCount keeps tracking of how many outstanding lifecycle termination
+	// events are being handled for Elasticsearch master-eligible nodes.
+	// Each goroutine handling the termination of a master-eligible node
+	// increments this this by 1 at the start using the `sync/atomic` package,
+	// and decrements by one after the lifecycle event completes.
+	// If the goroutine is the one that decrements down to 0, it will clear the master
+	// voting exclusions list.
+	votingCount int32
```
Comment (alphachai) [inline cmd/lifecycler/main.go]: What happens if the lifecycler dies while handling a lifecycle hook event? When it starts up again, would it just allow another autoscaling event?

Comment (alphachai) [reply cmd/lifecycler/main.go]: Or would aws wait on some default timeout for the lifecycle hook?

Comment (alphachai) [reply cmd/lifecycler/main.go]: I presume the sqs messages are on the queue still but not visible.

File: cmd/cloudwatcher/metrics.go (hunk 17 above)
Comment (alphachai) [reply cmd/cloudwatcher/metrics.go]: Would it be fair to say using variadic here is a personal preference? Just curious.

File: cmd/lifecycler/main.go (hunk 19 above)
Comment (jtdoepke) [reply cmd/lifecycler/main.go]: The hooks are on a timeout anyway. What lifecycler does is wait until just before the timeout, check if the cluster is good to go, and if not extend the timeout some more.

As for lifecycler dying: https://github.com/mintel/elasticsearch-asg/blob/master/cmd/lifecycler/main.go#L220

File: cmd/cloudwatcher/metrics.go (hunk 17 above)
Comment (jtdoepke) [reply cmd/cloudwatcher/metrics.go]: Yeah.

File: pkg/str/str.go [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+}
+
+// Uniq returns a new slice containing only unique strings.
+// The order of the return value is undefined.
+func Uniq(strs ...string) []string {
+	m := make(map[string]struct{}, len(strs))
+	for _, s := range strs {
+		m[s] = struct{}{}
+	}
+	out := make([]string, 0, len(m))
```
Comment (alphachai) [inline pkg/str/str.go]: If I understand correctly, the first two allocate memory for a string slice with zero elements. What does the third parameter do here?

Comment (jtdoepke) [reply pkg/str/str.go]: The third arg makes sure the string slice is backed by an underlying array with a capacity of `len(m)`. Since I know what the final size of the slice will be at the end, this prevents any reallocation/copying when I do `out = append(out, s)` later. It's a minor performance improvement.

https://blog.golang.org/go-slices-usage-and-internals

Comment (alphachai) [reply pkg/str/str.go]: Makes sense. Thank you!

File: pkg/es/cat_shards.go [hunk 21]
```diff
@@ -0,0 +1,237 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	// Return operation response
+	var ret CatShardsResponse
+	if err := (&elastic.DefaultDecoder{}).Decode(res.Body, &ret); err != nil {
+		return nil, err
+	}
+	return ret, nil
```
Comment (alphachai) [inline pkg/es/cat_shards.go]: Could this function be abstracted away? It seems very generic. Although, it might make sense in Go in a way I'm unaware.

Comment (jtdoepke) [reply pkg/es/cat_shards.go]: I'm not entirely sure what you mean here? The `Do()` method? Abstracted how?

Everything in `pkg/es` follows the style of https://github.com/olivere/elastic. I'd like to contribute the rest of these if he ever gets around to merging my initial [pull request](https://github.com/olivere/elastic/pull/1101).

Comment (alphachai) [reply pkg/es/cat_shards.go]: Ah, that makes sense. 👍
